File: AbstractFactoryPattern/python/PizzaStore/IngredientFactory.py
import abc
from .Ingredient import *

class IngredientFactory(metaclass=abc.ABCMeta):
    @abc.abstractmethod
    def createDough(self) -> Dough:
        return NotImplemented
    
    @abc.abstractmethod
    def createCheese(self) -> Cheese:
        return NotImplemented
    
    @abc.abstractmethod
    def createPepperoni(self) -> Pepperoni:
        return NotImplemented
    

class IngredientFactoryOfStoreA(IngredientFactory):
    def createDough(self) -> Dough:
        return ThinCrustDough()
    
    def createCheese(self) -> Cheese:
        return ReggianoCheese()
    
    def createPepperoni(self) -> Pepperoni:
        return SlicedPepperoni()
    

class IngredientFactoryOfStoreB(IngredientFactory):
    def createDough(self) -> Dough:
        return ThickCrustDough()
    
    def createCheese(self) -> Cheese:
        return MozzarellaCheese()
    
    def createPepperoni(self) -> Pepperoni:
        return SlicedPepperoni()

File: AbstractFactoryPattern/cpp/src/ingredient_factory.cpp
#include "header/ingredient_factory.h"


shared_ptr<Dough> IngredientFactoryOfStoreA::createDough(){
    return make_shared<ThinCrustDough>();
}

shared_ptr<Cheese> IngredientFactoryOfStoreA::createCheese(){
    return make_shared<ReggianoCheese>();
}

shared_ptr<Pepperoni> IngredientFactoryOfStoreA::createPepperoni(){
    return make_shared<SlicedPepperoni>();
}

shared_ptr<Dough> IngredientFactoryOfStoreB::createDough(){
    return make_shared<ThickCrustDough>();
}

shared_ptr<Cheese> IngredientFactoryOfStoreB::createCheese(){
    return make_shared<MozzarellaCheese>();
}

shared_ptr<Pepperoni> IngredientFactoryOfStoreB::createPepperoni(){
    return make_shared<SlicedPepperoni>();
}